File: functions/getCutoffs.R
getCutoffs = function(apps=NULL,vacs=NULL,assigned=NULL,cutoffs=NULL,iterat=NULL){
  cat(paste0('\n        Get cutoffs'))
  
  apps[,score:=priority_profile+lottery_number]
  apps = merge(apps,assigned[,.(applicant_id,program_id,assigned=1)],
               by=c('applicant_id','program_id'),all.x=T)
  apps[is.na(assigned),assigned:=0]
  apps = apps[order(program_id,-score)]
  uc = apps[assigned == 1,.(upper_cutoff=min(score,na.rm=T)),by=.(program_id)]
  lc = apps[assigned == 0,.(lower_cutoff=max(score,na.rm=T)),by=.(program_id)]
  
  vacs = unique(vacs[,.(program_id)])
  vacs = merge(vacs,uc,all.x=T)
  vacs = merge(vacs,lc,all.x=T)
  vacs[,not_filled:=ifelse(is.na(upper_cutoff),T,F)]
  vacs[is.na(upper_cutoff),upper_cutoff:=1]
  vacs[is.na(lower_cutoff),lower_cutoff:=1]
  vacs = vacs[,.(iter=iterat,program_id,not_filled,lower_cutoff,upper_cutoff)]
  cutoffs = rbind(cutoffs,vacs)
  return(cutoffs)
}

File: src/getCutoffs.py
import pandas as pd

def getCutoffs(apps=None, vacs=None, assigned=None, cutoffs=None, iterat=None, hard_quota=False):
    if not hard_quota:
        apps['score'] = apps['priority_profile'] + apps['lottery_number']
        apps = apps.merge(assigned[['applicant_id', 'program_id', 'assigned']], on=['applicant_id', 'program_id'], how='left')
        apps['assigned'].fillna(0, inplace=True)
        apps = apps.sort_values(by=['program_id', -apps['score']])
        uc = apps[apps['assigned'] == 1].groupby('program_id').agg(upper_cutoff=('score', 'min'))
        lc = apps[apps['assigned'] == 0].groupby('program_id').agg(lower_cutoff=('score', 'max'))
        vacs = vacs[['program_id']].drop_duplicates()
        vacs = vacs.merge(uc, on='program_id', how='left').merge(lc, on='program_id', how='left')
        vacs['not_filled'] = vacs['upper_cutoff'].isna()
        vacs.fillna(1, inplace=True)
        vacs['iter'] = iterat
    else:
        # Similar process as above but considering quotas
        # Please adjust as needed for your specific quota-related logic
        pass
    cutoffs = pd.concat([cutoffs, vacs], ignore_index=True)
    return cutoffs
